### PR TITLE
PIM-5109: As Peter, I would like to choose the products attributes to export

### PR DESCRIPTION
**Pull request by momoss**
Add filter attributes in ProductFlatArrayProcessor for filtering product attributes you want to export, only back office part at the moment.

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Y
| Added Behats                      | Y
| Changelog updated                 | Y
| Review and 2 GTM                  | Y
| Micro Demo to the PO (Story only) | N
| Migration script                  | N
| Tech Doc                          | N

### CODE REVIEW

File: src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php [hunk 1]
```diff
@@ -85,13 +100,72 @@ public function process($product)
...
+            );
+        }
+
         $this->detacher->detach($product);
 
         return $data;
     }
 
     /**
+     * filter the properties that has to be exported based on a product and a list of properties
```
Comment (pierallard) [inline src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php]: Filter

File: src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php [hunk 2]
```diff
@@ -85,13 +100,72 @@ public function process($product)
...
+        return array_intersect_key($normalizedProduct, array_flip($filteredColumnList));
+    }
+
+    /**
+     * @param ProductInterface $product
+     * @param string           $attributeCode
+     *
+     * @return AttributeInterface|null
+     */
+    protected function getProductAttributeByCode(ProductInterface $product, $attributeCode)
```
Comment (pierallard) [inline src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php]: getAttributeByCode?

File: src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php [hunk 3]
```diff
@@ -85,13 +100,72 @@ public function process($product)
...
+    {
+        foreach ($product->getValues() as $value) {
+            if ($attributeCode === $value->getAttribute()->getCode()) {
```
Comment (pierallard) [inline src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php]: Yoda style, variable on the right side plz

File: src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php [hunk 4]
```diff
@@ -103,20 +177,31 @@ public function setStepExecution(StepExecution $stepExecution)
...
     {
         $parameters = $this->stepExecution->getJobParameters();
         $decimalSeparator = $parameters->get('decimalSeparator');
         $dateFormat = $parameters->get('dateFormat');
+        $columns = isset($parameters->get('filters')['structure']['attributes']) ?
+            $parameters->get('filters')['structure']['attributes'] : [];
+
+        if (isset($columns) && 0 !== count($columns)) {
+            $columns[] = $this->attributeRepository->getIdentifierCode();
+        }
```
Comment (pierallard) [inline src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php]: Weird part.

What do you think of 
```
$columns = [];
if (isset($parameters->get('filters')['structure']['attributes'])) {
  $columns = $parameters->get('filters')['structure']['attributes'];
  if (count($columns) > 0) {
    $columns[] = $this->attributeRepository->getIdentifierCode();
  }
}
```

(isset($columns) is useless btw)

File: src/Pim/Bundle/BaseConnectorBundle/spec/Processor/ProductToFlatArrayProcessorSpec.php [hunk 5]
```diff
@@ -116,10 +122,11 @@ function it_returns_flat_data_with_media(
...
                 [
-                    'scopeCode'         => 'foobar',
-                    'localeCodes'       => ['en_US'],
-                    'decimal_separator' => '.',
-                    'date_format'       => 'yyyy-MM-dd',
+                    'scopeCode'           => 'foobar',
+                    'localeCodes'         => ['en_US'],
+                    'decimal_separator'   => '.',
+                    'date_format'         => 'yyyy-MM-dd',
+                    'selected_attributes' => [],
```
Comment (pierallard) [inline src/Pim/Bundle/BaseConnectorBundle/spec/Processor/ProductToFlatArrayProcessorSpec.php]: You added selected_attributes but not spec it!

File: src/Pim/Bundle/EnrichBundle/Connector/Processor/QuickExport/ProductToFlatArrayProcessor.php [hunk 6]
```diff
@@ -221,7 +230,7 @@ protected function getNormalizerContext()
         }
 
         if (isset($columns) && 0 !== count($columns)) {
-            $columns[] = 'sku';
+            $columns[] = $this->attributeRepository->getIdentifier();
```
Comment (pierallard) [inline src/Pim/Bundle/EnrichBundle/Connector/Processor/QuickExport/ProductToFlatArrayProcessor.php]: :heart:

File: src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php [hunk 7]
```diff
@@ -39,24 +42,36 @@ class ProductToFlatArrayProcessor extends AbstractConfigurableStepElement implem
...
+     * @param AttributeRepositoryInterface $attributeRepository
+     * @param string[]                     $mediaAttributeTypes
      */
     public function __construct(
         Serializer $serializer,
         ChannelRepositoryInterface $channelRepository,
         ProductBuilderInterface $productBuilder,
         ObjectDetacherInterface $detacher,
+        FieldSplitter $fieldSplitter,
+        AttributeRepositoryInterface $attributeRepository,
```
Comment (pierallard) [inline src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php]: CHANGELOG

File: src/Pim/Bundle/EnrichBundle/Connector/Processor/QuickExport/ProductToFlatArrayProcessor.php [hunk 8]
```diff
@@ -74,16 +79,18 @@ public function __construct(
         UserProviderInterface $userProvider,
         TokenStorageInterface $tokenStorage,
         FieldSplitter $fieldSplitter,
+        AttributeRepositoryInterface $attributeRepository,
```
Comment (pierallard) [inline src/Pim/Bundle/EnrichBundle/Connector/Processor/QuickExport/ProductToFlatArrayProcessor.php]: CHANGELOG

File: src/Pim/Bundle/EnrichBundle/spec/Connector/Processor/QuickExport/ProductToFlatArrayProcessorSpec.php [hunk 9]
```diff
@@ -47,6 +49,7 @@ function let(
             $userProvider,
             $tokenStorage,
             $fieldSplitter,
+            $attributeRepository,
```
Comment (pierallard) [inline src/Pim/Bundle/EnrichBundle/spec/Connector/Processor/QuickExport/ProductToFlatArrayProcessorSpec.php]: The fact you added $attributeRepository and you have nowhere "$attributeRepository->getIdentifier()->willReturn('sku')" looks like you have not tested this part of the code ;)

File: src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php [hunk 10]
```diff
@@ -85,13 +100,72 @@ public function process($product)
...
+            $field = $this->fieldSplitter->splitFieldName($key);
+            $attribute = $this->getProductAttributeByCode($product, $field[0]);
+
+            if (null === $attribute) {
+                $filteredColumnList[] = $key;
+            }
+
+            if (null !== $attribute && in_array($attribute->getCode(), $properties)) {
+                $filteredColumnList[] = $key;
+            }
```
Comment (pierallard) [inline src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php]: Replace this 2 conditions by this:
if (null === $attribute || in_array($attribute->getCode, $properties)

File: src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php [hunk 11]
```diff
@@ -85,13 +100,72 @@ public function process($product)
...
     /**
+     * filter the properties that has to be exported based on a product and a list of properties
+     *
+     * @param ProductInterface $product
+     * @param array            $normalizedProduct
+     * @param array            $properties
+     *
+     * @return array
+     */
+    protected function filterAttributes(
```
Comment (pierallard) [inline src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php]: function filterProductAttributes

File: src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php [hunk 12]
```diff
@@ -85,13 +100,72 @@ public function process($product)
...
+        ProductInterface $product,
+        array $normalizedProduct,
+        array $properties
+    ) {
+        $filteredColumnList = [];
```
Comment (pierallard) [inline src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php]: filteredAttributes

**Comment by pierallard** [comment]
Some little refacto and GTM

File: src/Pim/Bundle/EnrichBundle/Connector/Processor/QuickExport/ProductToFlatArrayProcessor.php (hunk 6 above)
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Connector/Processor/QuickExport/ProductToFlatArrayProcessor.php]: it return a string?

File: src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php (hunk 3 above)
Comment (momoss) [inline src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php]: Look at $product->getValue($attributeCode, $localeCode = null, $scopeCode = null);

File: src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php (hunk 4 above)
Comment (momoss) [inline src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php]: $columns = [];
if (isset($parameters->get('filters')['structure']['attributes'])) {
  $columns = $parameters->get('filters')['structure']['attributes'];
  if (!in_array($this->attributeRepository->getIdentifierCode(), $columns) {
    $columns[] = $this->attributeRepository->getIdentifierCode();
  }
}

File: src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php [hunk 13]
```diff
@@ -85,13 +100,72 @@ public function process($product)
             );
         }
 
-        $data['product'] = $this->serializer->normalize($product, 'flat', $this->getNormalizerContext($contextChannel));
+        $normalizerContext = $this->getNormalizerContext($contextChannel, $product);
+        $data['product'] = $this->serializer->normalize($product, 'flat', $normalizerContext);
+
```
Comment (momoss) [inline src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php]: Remove $columns check from normalizerContext

File: src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php [hunk 14]
```diff
@@ -103,20 +177,31 @@ public function setStepExecution(StepExecution $stepExecution)
...
-            'decimal_separator' => $decimalSeparator,
-            'date_format'       => $dateFormat,
+            'scopeCode'           => $channel->getCode(),
+            'localeCodes'         => array_intersect(
+                $channel->getLocaleCodes(),
+                $parameters->get('filters')['structure']['locales']
+            ),
+            'decimal_separator'   => $decimalSeparator,
+            'date_format'         => $dateFormat,
+            'selected_attributes' => $columns,
```
Comment (momoss) [inline src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php]: Remove line

File: src/Pim/Bundle/EnrichBundle/Connector/Processor/QuickExport/ProductToFlatArrayProcessor.php (hunk 6 above)
Comment (momoss) [inline src/Pim/Bundle/EnrichBundle/Connector/Processor/QuickExport/ProductToFlatArrayProcessor.php]: ->getCode()

File: features/export/xlsx/export_products.feature [hunk 15]
```diff
@@ -96,3 +96,65 @@ Feature: Export products in XLSX
...
+    And I launched the completeness calculator
+    When I am on the "xlsx_tablet_product_export" export job page
+    And I launch the export job
+    And I wait for the "xlsx_tablet_product_export" job to finish
+    And exported xlsx file of "xlsx_tablet_product_export" should contain:
+      | sku           | categories                 |  enabled | family  | groups |
+      | tshirt-white  | men_2013,men_2014,men_2015 |  1       | tshirts |        |
+      | tshirt-black  | men_2013,men_2014,men_2015 |  1       | tshirts |        |
+      | tshirt-yellow | men_2013,men_2014,men_2015 |  1       | tshirts |        |
+      | tshirt-green  | men_2013,men_2014,men_2015 |  1       | tshirts |        |
```
Comment (juliensnz) [inline features/export/xlsx/export_products.feature]: is there a new line at the end of the file?

File: features/export/xlsx/export_products.feature [hunk 16]
```diff
@@ -96,3 +96,65 @@ Feature: Export products in XLSX
...
       | tshirt-yellow |                   | men_2013,men_2014,men_2015 | yellow | 10.00    | 20.00    | 30.00    | usa                    | 2                      |           |                          | A stylish yellow t-shirt | 1       | tshirts |        | 1        | files/tshirt-yellow/image/SNKRS-1R.png |              |              | american_apparel | cotton   | Yellow t-shirt | Yellow t-shirt | 10                     | 10.00     | 9.00      | 15.00     | 2016-10-12          | size_M |           |                     |                          | 5      | KILOGRAM    |
       | tshirt-green  |                   | men_2013,men_2014,men_2015 | green  |          |          |          | usa                    |                        |           |                          |                          | 1       | tshirts |        |          |                                        |              |              | american_apparel | cotton   | Green t-shirt  | Green t-shirt  |                        | 10.00     | 9.00      | 15.00     |                     | size_L |           |                     |                          |        |             |
+
+  @javascript
```
Comment (juliensnz) [inline features/export/xlsx/export_products.feature]: you don't test it with csv?

File: features/export/xlsx/export_products.feature (hunk 15 above)
Comment (momoss) [inline features/export/xlsx/export_products.feature]: Added :-)

File: features/export/xlsx/export_products.feature (hunk 16 above)
Comment (momoss) [inline features/export/xlsx/export_products.feature]: Added 😄

File: src/Pim/Bundle/BaseConnectorBundle/spec/Processor/ProductToFlatArrayProcessorSpec.php [hunk 17]
```diff
@@ -116,10 +125,10 @@ function it_returns_flat_data_with_media(
                 $product,
                 'flat',
                 [
-                    'scopeCode'         => 'foobar',
-                    'localeCodes'       => ['en_US'],
-                    'decimal_separator' => '.',
-                    'date_format'       => 'yyyy-MM-dd',
+                    'scopeCode'          => 'foobar',
+                    'localeCodes'        => ['en_US'],
+                    'decimal_separator'  => '.',
```
Comment (juliensnz) [inline src/Pim/Bundle/BaseConnectorBundle/spec/Processor/ProductToFlatArrayProcessorSpec.php]: why?

File: src/Pim/Bundle/BaseConnectorBundle/spec/Processor/ProductToFlatArrayProcessorSpec.php [hunk 18]
```diff
@@ -170,15 +180,17 @@ function it_returns_flat_data_without_media(
...
                 $product,
                 'flat',
                 [
-                    'scopeCode'         => 'mobile',
-                    'localeCodes'       => ['en_US'],
-                    'decimal_separator' => ',',
-                    'date_format'       => 'yyyy-MM-dd',
+                    'scopeCode'          => 'mobile',
+                    'localeCodes'        => ['en_US'],
+                    'decimal_separator'  => ',',
```
Comment (juliensnz) [inline src/Pim/Bundle/BaseConnectorBundle/spec/Processor/ProductToFlatArrayProcessorSpec.php]: why?

**Comment by juliensnz** [comment]
GTM after comment fix

File: features/export/export_products.feature [hunk 19]
```diff
@@ -240,3 +240,93 @@ Feature: Export products
...
+      | tshirt-yellow | description     | A really stylish yellow t-shirt      | en_US  | print     |
+      | tshirt-green  | name            | Green t-shirt                        | en_US  |           |
+      | tshirt-green  | name            | Green t-shirt                        | en_GB  |           |
+      | tshirt-green  | name            | T-shirt noir                         | fr_FR  |           |
+      | tshirt-green  | name            | Schwarzes T-Shirt                    | de_DE  |           |
+      | tshirt-green  | description     | Un T-shirt noir élégant              | fr_FR  | ecommerce |
+      | tshirt-green  | description     | Ein elegantes schwarzes T-Shirt      | de_DE  | ecommerce |
+      | tshirt-green  | description     | A really stylish green t-shirt       | en_US  | print     |
+      | tshirt-green  | description     | Ein sehr elegantes schwarzes T-Shirt | de_DE  | print     |
+    And I launched the completeness calculator
```
Comment (phaseinducer) [inline features/export/export_products.feature]: Not sure it's still useful

File: features/export/export_products.feature [hunk 20]
```diff
@@ -240,3 +240,93 @@ Feature: Export products
...
+    And the following products:
+      | sku           | family  | categories                   | price                 | size   | color  | manufacturer     | material | country_of_manufacture |
+      | tshirt-yellow | tshirts | men_2013, men_2014, men_2015 | 10 EUR, 15 USD, 9 GBP | size_M | yellow | american_apparel | cotton   | usa                    |
+      | tshirt-green  | tshirts | men_2013, men_2014, men_2015 | 10 EUR, 15 USD, 9 GBP | size_L | green  | american_apparel | cotton   | usa                    |
+    And the following product values:
+      | product       | attribute       | value                                | locale | scope     |
+      | tshirt-yellow | name            | Yellow t-shirt                       | en_US  |           |
+      | tshirt-yellow | name            | Yellow t-shirt                       | en_GB  |           |
+      | tshirt-yellow | name            | T-shirt blanc                        | fr_FR  |           |
+      | tshirt-yellow | name            | Weißes T-Shirt                       | de_DE  |           |
```
Comment (phaseinducer) [inline features/export/export_products.feature]: Weisses

File: features/export/export_products.feature [hunk 21]
```diff
@@ -240,3 +240,93 @@ Feature: Export products
...
+    And I launch the export job
+    And I wait for the "tablet_product_export" job to finish
+    Then exported file of "tablet_product_export" should contain:
+    """
+    sku;categories;enabled;family;groups
+    tshirt-yellow;men_2013,men_2014,men_2015;1;tshirts;
+    tshirt-green;men_2013,men_2014,men_2015;1;tshirts;
+    """
+
+  Scenario: Successfully export products with associations and an empty selection of attributes
```
Comment (phaseinducer) [inline features/export/export_products.feature]: Why test it?

File: features/export/export_products.feature [hunk 22]
```diff
@@ -240,3 +240,93 @@ Feature: Export products
...
+    sku;categories;color;cost-EUR;cost-GBP;cost-USD;description-en_GB-tablet;description-en_US-tablet;enabled;family;groups;image;name-en_GB;name-en_US;price-EUR;price-GBP;price-USD;release_date-tablet;size;weight;weight-unit
+    tshirt-yellow;men_2013,men_2014,men_2015;yellow;10.00;20.00;30.00;;A stylish yellow t-shirt;1;tshirts;;files/tshirt-yellow/image/SNKRS-1R.png;Yellow t-shirt;Yellow t-shirt;10.00;9.00;15.00;2016-10-12;size_M;5;KILOGRAM
+    tshirt-green;men_2013,men_2014,men_2015;green;;;;;;1;tshirts;;;Green t-shirt;Green t-shirt;10.00;9.00;15.00;;size_L;;
+    """
+
+  Scenario: Successfully export products with an empty selection of attributes
+    Given an "apparel" catalog configuration
+    And I am logged in as "Julia"
+    And the following job "tablet_product_export" configuration:
+      | filters  | {"structure":{"locales":["en_US","fr_FR"],"scope":"tablet","attributes":[]}, "data": []} |
```
Comment (phaseinducer) [inline features/export/export_products.feature]: extra space

File: features/export/xlsx/export_products.feature [hunk 23]
```diff
@@ -96,3 +96,88 @@ Feature: Export products in XLSX
...
+  Scenario: Successfully export products with a selection of attributes
+    Given the following job "xlsx_tablet_product_export" configuration:
+      | filters  | {"structure":{"locales":["en_US","fr_FR"],"scope":"tablet","attributes":["price","size","color","cost","description","name","image","release_date","weight"]}, "data": []} |
```
Comment (phaseinducer) [inline features/export/xlsx/export_products.feature]: extra space

File: features/export/xlsx/export_products.feature [hunk 24]
```diff
@@ -96,3 +96,88 @@ Feature: Export products in XLSX
...
+      | tshirt-black  | men_2013,men_2014,men_2015 | black  |          |          |          |                          |                          | 1       | tshirts |        |                                        | Black t-shirt  | Black t-shirt  | 10.00     | 9.00      | 15.00     |                     |        |        |             |
+      | tshirt-yellow | men_2013,men_2014,men_2015 | yellow | 10.00    | 20.00    | 30.00    |                          | A stylish yellow t-shirt | 1       | tshirts |        | files/tshirt-yellow/image/SNKRS-1R.png | Yellow t-shirt | Yellow t-shirt | 10.00     | 9.00      | 15.00     | 2016-10-12          | size_M | 5      | KILOGRAM    |
+      | tshirt-green  | men_2013,men_2014,men_2015 | green  |          |          |          |                          |                          | 1       | tshirts |        |                                        | Green t-shirt  | Green t-shirt  | 10.00     | 9.00      | 15.00     |                     | size_L |        |             |
+
+  @javascript
```
Comment (phaseinducer) [inline features/export/xlsx/export_products.feature]: You can remove it everywhere and put it on the feature.

File: src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php [hunk 25]
```diff
@@ -86,12 +101,86 @@ public function process($product)
...
+     * @return array|null
+     */
+    protected function getFilterableAttributes()
+    {
+        $attributes = null;
+        $parameters = $this->stepExecution->getJobParameters();
+
+        if (isset($parameters->get('filters')['structure']['attributes'])) {
+            $attributes = $parameters->get('filters')['structure']['attributes'];
+            $attributeCode = $this->attributeRepository->getIdentifierCode();
```
Comment (phaseinducer) [inline src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php]: `$identifierCode`

File: src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php [hunk 26]
```diff
@@ -86,12 +101,86 @@ public function process($product)
         }
 
         $data['product'] = $this->serializer->normalize($product, 'flat', $this->getNormalizerContext($contextChannel));
+
+        $attributes = $this->getFilterableAttributes();
+        if (null !== $attributes) {
```
Comment (phaseinducer) [inline src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php]: What's the case for `null` here?

File: src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php [hunk 27]
```diff
@@ -86,12 +101,86 @@ public function process($product)
...
+
         $this->detacher->detach($product);
 
         return $data;
     }
 
     /**
+     * @return array|null
+     */
+    protected function getFilterableAttributes()
```
Comment (phaseinducer) [inline src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php]: Naming problem imho, maybe `getAttributesToFilter`?

File: src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php [hunk 28]
```diff
@@ -86,12 +101,86 @@ public function process($product)
...
+            if (!in_array($attributeCode, $attributes)) {
+                $attributes[] = $attributeCode;
+            }
+        }
+
+        return $attributes;
+    }
+
+    /**
+     * filters the attributes that have to be exported based on a product and a list of attributes
```
Comment (phaseinducer) [inline src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php]: Filters

File: features/export/export_products.feature [hunk 29]
```diff
@@ -240,3 +240,93 @@ Feature: Export products
...
+      | sku           | family  | categories                   | price                 | size   | color  | manufacturer     | material | country_of_manufacture |
+      | tshirt-yellow | tshirts | men_2013, men_2014, men_2015 | 10 EUR, 15 USD, 9 GBP | size_M | yellow | american_apparel | cotton   | usa                    |
+      | tshirt-green  | tshirts | men_2013, men_2014, men_2015 | 10 EUR, 15 USD, 9 GBP | size_L | green  | american_apparel | cotton   | usa                    |
+    And I launched the completeness calculator
+    When I am on the "tablet_product_export" export job page
+    And I launch the export job
+    And I wait for the "tablet_product_export" job to finish
+    Then exported file of "tablet_product_export" should contain:
+    """
+    sku;categories;enabled;family;groups
```
Comment (phaseinducer) [inline features/export/export_products.feature]: With `"attributes":[]` you should have all attributes exported.

File: src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php [hunk 30]
```diff
@@ -86,12 +101,86 @@ public function process($product)
...
+     * @return array
+     */
+    protected function filterAttributes(
+        ProductInterface $product,
+        array $normalizedProduct,
+        array $attributes
+    ) {
+        $filteredColumnList = [];
+
+        foreach ($normalizedProduct as $key => $value) {
```
Comment (phaseinducer) [inline src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php]: You don't use `$value` so you should do something like `foreach (array_keys($normalizedProduct) as $flatField) {` (btw if you don't you'll get a pmd warning)

File: src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php [hunk 31]
```diff
@@ -86,12 +101,86 @@ public function process($product)
...
+            $field = $this->fieldSplitter->splitFieldName($key);
```
Comment (phaseinducer) [inline src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php]: `$fieldParts`

File: src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php [hunk 32]
```diff
@@ -86,12 +101,86 @@ public function process($product)
...
+
+    /**
+     * @param ProductInterface $product
+     * @param string           $attributeCode
+     *
+     * @return AttributeInterface|null
+     */
+    protected function getAttributeByCode(ProductInterface $product, $attributeCode)
+    {
+        foreach ($product->getValues() as $value) {
```
Comment (phaseinducer) [inline src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php]: It means that for each column, you loop on all product values, it's a bit heavy. You should do this only once by product (you can do it in the `initialize()`).

File: src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php [hunk 33]
```diff
@@ -113,10 +202,13 @@ protected function getNormalizerContext(ChannelInterface $channel)
...
-            'localeCodes'       => array_intersect($channel->getLocaleCodes(), $parameters->get('filters')['structure']['locales']),
-            'decimal_separator' => $decimalSeparator,
-            'date_format'       => $dateFormat,
+            'scopeCode'           => $channel->getCode(),
+            'localeCodes'         => array_intersect(
+                $channel->getLocaleCodes(),
+                $parameters->get('filters')['structure']['locales']
+            ),
+            'decimal_separator'   => $decimalSeparator,
+            'date_format'         => $dateFormat,
```
Comment (phaseinducer) [inline src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php]: You can revert this part

File: src/Pim/Bundle/EnrichBundle/Connector/Processor/QuickExport/ProductToFlatArrayProcessor.php [hunk 34]
```diff
@@ -221,7 +228,7 @@ protected function getNormalizerContext()
         }
 
         if (isset($columns) && 0 !== count($columns)) {
-            $columns[] = 'sku';
+            $columns[] = $this->attributeRepository->getIdentifierCode();
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Connector/Processor/QuickExport/ProductToFlatArrayProcessor.php]: gg

File: src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php [hunk 35]
```diff
@@ -86,12 +101,86 @@ public function process($product)
...
+            );
+        }
+
         $this->detacher->detach($product);
 
         return $data;
     }
 
     /**
+     * @return array|null
```
Comment (phaseinducer) [inline src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php]: A little description is needed imho

**Comment by phaseinducer** [comment]
GTM

**Pull request merged**